torch/csrc/jit/codegen/cuda/contiguity.cpp

```diff
@@ -8,20 +9,455 @@ namespace jit {
+
+  // Update maps
+  // Find the position inner would have to have to be considered ordered
+  auto pos_after_outer = outer_pos + 1;
+  for (; pos_after_outer < active_ids_.size(); pos_after_outer++) {
+    if (active_ids_[pos_after_outer] == nullptr) {
+      // Can't be considered ordered after a nullptr
+      break;
+    }
+    if (active_ids_[pos_after_outer]->isReduction() ||
+        ((active_ids_[pos_after_outer]->isBroadcast() &&
+          !concrete_info_->isConcretized(active_ids_[pos_after_outer])))) {
+      // Skip reduction or broadcast axes that aren't concretized in the fusion
+      continue;
+    }
+    break;
+  }
+
+  // The output is ordered as long as the inputs were ordered and outer position
+  // is directly left of the inner position.
+  bool out_ordered = inner_ordered && outer_ordered;
+  out_ordered = out_ordered &&
+      // If inner_pos is before outer_pos it's not ordered correctly. If for
+      // some reason it's the same, that would be an error.
+      inner_pos > outer_pos &&
+      // Inner could be a broadcast, so doesn't have to be right on
+      // pos_after_outer as that ID (if it exists) should not be a broadcast.
+      // However, merging over a broadcast should be fine.
+      inner_pos <= pos_after_outer && !inner_is_concretized_bcast;
+
+  if (out_ordered) {
+    consistently_ordered_ids_.emplace(merge->out());
+  }
+
+  // Don't just remove active_ids_, as if we have something like:
+  //   [i0, i1, i2, i3]
+  //   ->merge(0, 2)
+  //   ->merge(1)
+  // The latter merge looks like it's ordered correctly, if we update the active
+  // map as:
+  //   [i0, i1, i2, i3] -> [i0*i2, i1, i3]
+  // Hoever if we instead mark it as:
+  //   [i0, i1, i2, i3] -> [i0*i2, i1, nullptr, i3]
+  // Or:
+  //   [i0, i1, i2, i3] -> [nullptr, i1, i0*i2, i3]
+  // It's clear the second merge is not ordered correctly. Doesn't matter which
+  // direction we put the iter domain in, prefer putting it in outer as we often
+  // are looking for inner dimensions that are contiguous. We don't want to
+  // always do this, as it could make ordered merges look non-ordered.
+  // For exmaple: [i0, i1, i2, i3]
+  // ->merge(0)
+  // ->merge(1)
+  // ->merge(0)
+  // If it's updated as:
+  // [i0, i1, i2, i3]
+  // -> [i0*i1, nullptr, i2, i3]
+  // -> [i0*i1, nullptr, i2*i3, nullptr]
+  // Now the final merge looks non-ordered but it is. So only insert a nullptr
+  // entry if the out is not ordered.
+  active_ids_[outer_pos] = merge->out();
+
+  if (!out_ordered) {
+    active_ids_[inner_pos] = nullptr;
+  } else {
+    active_ids_.erase(active_ids_.begin() + inner_pos);
+    for (auto i = outer_pos + 1; i < inner_pos; i++) {
+      // If there's broadcast axes between outer and inner and the merge was
+      // contiguous, there may be broadcasts between outer and inner that cannot
+      // be ordered merged anywhere else so remove them.
+      active_ids_.erase(active_ids_.begin() + outer_pos + 1);
```

Comment:
Is this necessary? Or is it just for efficiency?

Reply:
I think necessary:
```
T0[i0, b1, i2, i3]
T0->merge(0, 2)->merge(1,2)->merge(0)
```
Without erase active domains would play forward as:
```
0: [i0, b1, i2, i3]
1: [i0*i2, b1, i3]
2: [i0*i2, b1*i3]
3: [i0*i2*b1*i3]
```
So even though stage 1 is correctly ordered, stage  2 is not, so we want to make sure we don't incorrectly view this as ordered.

Reply:
I see. Why not just removing all non-concretized broadcasts and reductions first?

Reply:
I think because I was hoping we would use local information not global information on broadcast resolution. I should only consider it non-contiguous merge across the broadcast if that merge is coming from something that was considering it concretized. So like in the transpose scheduler when we have two groups it'd be nice if they were considered independently of eachother, and the over all approach made here prevents separation of concerns.

Given where it's at I agree, we should probably do that, and process it as it's just a domain that interferes.

Reply:
Maybe I wasn't clear. What I suggested was just remove all reduction and non-concretized broadcast domains from `active_ids_` in the constructor before traversing down the expressions. This way, we could remove this logic as well: https://github.com/csarofeen/pytorch/pull/1990/files/8bd625c7373e6eab891e2d51d099b137644e3e18#diff-58f78ed0235f3bc2f291e26946b8713607ebe37099e0a0e9395ace85c293365aR137-R142.

This is just a refactoring suggestion.

Reply:
We have to remove the ids as we update `active_ids_`, so it's hard to change this part of the logic since we only really have this in the pass checking the ordering of the operations. When we can traverse directly on ID graphs I think this could be done more straightforwardly, but if there's a way to check this property when not traversing/updating `active_ids_` it eludes me. Leaving as is for now, but also going to check if outer is a concretized bcast.